[question] CMakeToolchain requires VCVars

### What is your question?

We distribute build tools (including msvc, and the SDKs),  internally in our company via `build_requires` in our build profiles

For example: 

```
[settings]
os=Windows
arch=x86_64
compiler=msvc
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=192
compiler.update=8
compiler.cppstd=14
build_type=Release

[options]
[build_requires]
cmake/3.24.2
ninja/1.11.1
vc/19.28@internal/dk
ms_sdk/21322@internal/dk
ms_wdk/21322@internal/dk

[env]
CMAKE_BUILD_TYPE=Release
CMAKE_GENERATOR=Ninja

```
the `vc`, `ms_sdk`, and `ms_wdk` recipes were created by us.  They package up the build tool assets (binaries, headers, resource files, etc...), sets environment variables similar to what `vcvarsall.bat` does.  

This allows us to quickly change compilers with a simple build profile change.

With `CMakeToolChain`,  it is expecting to run vcvars for the `msvc` compiler,  which typically doesn't exist in our environments.

Before I go too deep into describing our situation,  is there a proper way to support msvc build tools via build requires?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

Thanks for the info, I'll be sure to update our recipes.

It would be nice if we could avoid VCVars all together

Here's an interesting data point.  If I change my `settings.yml` to copy the `msvc` compiler entry to `msvc_custom` .  and use that instead,  I am able to build the zlib package

```
Z:\repos>conan install zlib/1.2.11@ -pr release_msvc_19.28 --build=zlib
Configuration:
[settings]
arch=x86_64
build_type=Release
compiler=msvc_custom
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=192
os=Windows
[options]
[build_requires]
*: cmake/3.24.2, ninja/1.11.1, vc/19.28@internal/dk, ms_sdk/21322@internal/dk, ms_wdk/21322@internal/dk
[env]
CMAKE_BUILD_TYPE=Release
CMAKE_GENERATOR=Ninja
[conf]
tools.cmake.cmaketoolchain:generator=Ninja

zlib/1.2.11: Forced build from source
cmake/3.24.2: WARN: cmake recipe lacks information about the msvc_custom compiler standard version support
cmake/3.24.2: WARN: cmake requires a compiler that supports at least C++11
Installing package: zlib/1.2.11
Requirements
    zlib/1.2.11 from 'conancenter' - Cache
Packages
    zlib/1.2.11:7729185d5890c7a353e9f33596c2f8f5960e65c8 - Build
Build requirements
    cmake/3.24.2 from 'conancenter' - Cache
    ms_sdk/21322@internal/dk from 'internal_conan_local' - Cache
    ms_wdk/21322@internal/dk from 'internal_conan_local' - Cache
    ninja/1.11.1 from 'conancenter' - Cache
    vc/19.28@internal/dk from 'internal_conan_local' - Cache
Build requirements packages
    cmake/3.24.2:0a420ff5c47119e668867cdb51baff0eca1fdb68 - Cache
    ms_sdk/21322@internal/dk:ca33edce272a279b24f87dc0d4cf5bbdcffbc187 - Cache
    ms_wdk/21322@internal/dk:ca33edce272a279b24f87dc0d4cf5bbdcffbc187 - Cache
    ninja/1.11.1:0a420ff5c47119e668867cdb51baff0eca1fdb68 - Cache
    vc/19.28@internal/dk:ca33edce272a279b24f87dc0d4cf5bbdcffbc187 - Cache

Installing (downloading, building) binaries...
cmake/3.24.2: Already installed!
cmake/3.24.2: Appending PATH environment variable: C:\Users\nschultz\.conan\data\cmake\3.24.2\_\_\package\0a420ff5c47119e668867cdb51baff0eca1fdb68\bin
ms_sdk/21322@internal/dk: Already installed!
ms_wdk/21322@internal/dk: Already installed!
ninja/1.11.1: Already installed!
vc/19.28@internal/dk: Already installed!
zlib/1.2.11: Applying build-requirement: cmake/3.24.2
zlib/1.2.11: Applying build-requirement: ninja/1.11.1
zlib/1.2.11: Applying build-requirement: vc/19.28@internal/dk
zlib/1.2.11: Applying build-requirement: ms_sdk/21322@internal/dk
zlib/1.2.11: Applying build-requirement: ms_wdk/21322@internal/dk
zlib/1.2.11: Copying sources to build folder
zlib/1.2.11: Building your package in C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8
zlib/1.2.11: Generator txt created conanbuildinfo.txt
zlib/1.2.11: Calling generate()
zlib/1.2.11: WARN: Using the new toolchains and generators without specifying a build profile (e.g: -pr:b=default) is discouraged and might cause failures and unexpected behavior
zlib/1.2.11: Preset 'release' added to CMakePresets.json. Invoke it manually using 'cmake --preset release'
zlib/1.2.11: If your CMake version is not compatible with CMakePresets (<3.19) call cmake like: 'cmake <path> -G Ninja -DCMAKE_TOOLCHAIN_FILE=C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\build\generators\conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release'
zlib/1.2.11: Aggregating env generators
zlib/1.2.11: Calling build()
zlib/1.2.11: CMake command: cmake -G "Ninja" -DCMAKE_TOOLCHAIN_FILE="C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/build/7729185d5890c7a353e9f33596c2f8f5960e65c8/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/package/7729185d5890c7a353e9f33596c2f8f5960e65c8" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\src"
-- Using Conan toolchain: C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/build/7729185d5890c7a353e9f33596c2f8f5960e65c8/build/generators/conan_toolchain.cmake
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The C compiler identification is MSVC 19.28.29910.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Users/nschultz/.conan/data/vc/19.28/internal/dk/package/ca33edce272a279b24f87dc0d4cf5bbdcffbc187/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of off64_t
-- Check size of off64_t - failed
-- Looking for fseeko
-- Looking for fseeko - not found
-- Looking for unistd.h
-- Looking for unistd.h - not found
-- Renaming
--     C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/build/7729185d5890c7a353e9f33596c2f8f5960e65c8/src/zconf.h
-- to 'zconf.h.included' because this file is included with zlib
-- but CMake generates it automatically in the build directory.
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/build/7729185d5890c7a353e9f33596c2f8f5960e65c8/build/Release
zlib/1.2.11: CMake command: cmake --build "C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\build\Release" -- -j8
[9/16] Building C object CMakeFiles\zlib.dir\gzwrite.c.obj
C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\src\gzwrite.c(212): warning C4267: '=': conversion from 'size_t' to 'unsigned int', possible loss of data
C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\src\gzwrite.c(232): warning C4267: '=': conversion from 'size_t' to 'unsigned int', possible loss of data
C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\src\gzwrite.c(371): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
[11/16] Building C object CMakeFiles\zlib.dir\gzread.c.obj
C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\src\gzread.c(319): warning C4267: '=': conversion from 'size_t' to 'unsigned int', possible loss of data
C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\src\gzread.c(400): warning C4267: '=': conversion from 'size_t' to 'unsigned int', possible loss of data
C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\src\gzread.c(472): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
[16/16] Linking C static library zlib.lib
zlib/1.2.11: Package '7729185d5890c7a353e9f33596c2f8f5960e65c8' built
zlib/1.2.11: Build folder C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\build\Release
zlib/1.2.11: Generated conaninfo.txt
zlib/1.2.11: Generated conanbuildinfo.txt
zlib/1.2.11: Generating the package
zlib/1.2.11: Package folder C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\package\7729185d5890c7a353e9f33596c2f8f5960e65c8
zlib/1.2.11: Calling package()
zlib/1.2.11: CMake command: cmake --install "C:\Users\nschultz\.conan\data\zlib\1.2.11\_\_\build\7729185d5890c7a353e9f33596c2f8f5960e65c8\build\Release" --prefix "C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/package/7729185d5890c7a353e9f33596c2f8f5960e65c8"
-- Install configuration: "Release"
-- Installing: C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/package/7729185d5890c7a353e9f33596c2f8f5960e65c8/lib/zlib.lib
-- Installing: C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/package/7729185d5890c7a353e9f33596c2f8f5960e65c8/include/zconf.h
-- Installing: C:/Users/nschultz/.conan/data/zlib/1.2.11/_/_/package/7729185d5890c7a353e9f33596c2f8f5960e65c8/include/zlib.h
zlib/1.2.11 package(): Packaged 2 '.h' files: zconf.h, zlib.h
zlib/1.2.11 package(): Packaged 1 '.lib' file: zlib.lib
zlib/1.2.11 package(): Packaged 1 file: LICENSE
zlib/1.2.11: Package '7729185d5890c7a353e9f33596c2f8f5960e65c8' created
zlib/1.2.11: Created package revision 482d14c1444c34f40ef9ebe62aa27aaf
Aggregating env generators
```

--- a/conan/tools/microsoft/visual.py
+++ b/conan/tools/microsoft/visual.py
@@ -73,7 +73,7 @@ class VCVars:
             return
 
         vs_install_path = conanfile.conf.get("tools.microsoft.msbuild:installation_path")
-        if vs_install_path == "disabled":
+        if vs_install_path == "":  # Empty string means "disable"
             return
 
         if compiler == "clang":

--- a/conan/tools/microsoft/visual.py
+++ b/conan/tools/microsoft/visual.py
@@ -72,6 +72,10 @@ class VCVars:
         if compiler not in ("Visual Studio", "msvc", "clang"):
             return
 
+        vs_install_path = conanfile.conf.get("tools.microsoft.msbuild:installation_path")
+        if vs_install_path == "disabled":
+            return
+
         if compiler == "clang":
             # The vcvars only needed for LLVM/Clang and VS ClangCL, who define runtime
             if not conanfile.settings.get_safe("compiler.runtime"):
@@ -93,7 +97,6 @@ class VCVars:
             vcvars_ver = _vcvars_vers(conanfile, compiler, vs_version)
         vcvarsarch = vcvars_arch(conanfile)
 
-        vs_install_path = conanfile.conf.get("tools.microsoft.msbuild:installation_path")
         # The vs_install_path is like
         # C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
         # C:\Program Files (x86)\Microsoft Visual Studio\2017\Community

--- a/conans/model/conf.py
+++ b/conans/model/conf.py
@@ -35,7 +35,7 @@ BUILT_IN_CONFS = {
     "tools.microsoft.msbuild:verbosity": "Verbosity level for MSBuild: 'Quiet', 'Minimal', 'Normal', 'Detailed', 'Diagnostic'",
     "tools.microsoft.msbuild:vs_version": "Defines the IDE version when using the new msvc compiler",
     "tools.microsoft.msbuild:max_cpu_count": "Argument for the /m when running msvc to build parallel projects",
-    "tools.microsoft.msbuild:installation_path": "VS install path, to avoid auto-detect via vswhere, like C:/Program Files (x86)/Microsoft Visual Studio/2019/Community",
+    "tools.microsoft.msbuild:installation_path": "VS install path, to avoid auto-detect via vswhere, like C:/Program Files (x86)/Microsoft Visual Studio/2019/Community. Use empty string to disable",
     "tools.microsoft.msbuilddeps:exclude_code_analysis": "Suppress MSBuild code analysis for patterns",
     "tools.microsoft.msbuildtoolchain:compile_options": "Dictionary with MSBuild compiler options",
     "tools.intel:installation_path": "Defines the Intel oneAPI installation root path",

--- a/conans/test/integration/toolchains/microsoft/vcvars_test.py
+++ b/conans/test/integration/toolchains/microsoft/vcvars_test.py
@@ -45,8 +45,12 @@ def test_vcvars_generator_skip():
     client = TestClient()
     client.save({"conanfile.py": GenConanfile().with_generator("VCVars")
                                                .with_settings("os", "compiler",
-                                                              "arch", "build_type")})
-    client.run('install . -c tools.microsoft.msbuild:installation_path=disabled')
+                                                              "arch", "build_type"),
+                 "profile": 'include(default)\n[conf]\ntools.microsoft.msbuild:installation_path='})
+
+    client.run('install . -c tools.microsoft.msbuild:installation_path=""')
+    assert not os.path.exists(os.path.join(client.current_folder, "conanvcvars.bat"))
+    client.run('install . -pr=profile')
     assert not os.path.exists(os.path.join(client.current_folder, "conanvcvars.bat"))
 
 

--- a/conans/test/integration/toolchains/microsoft/vcvars_test.py
+++ b/conans/test/integration/toolchains/microsoft/vcvars_test.py
@@ -4,6 +4,7 @@ import os
 
 import pytest
 
+from conans.test.assets.genconanfile import GenConanfile
 from conans.test.utils.tools import TestClient
 
 
@@ -34,6 +35,19 @@ def test_vcvars_generator(scope):
         assert "conanvcvars.bat" in bat_contents
     else:
         assert not os.path.exists(os.path.join(client.current_folder, "conanbuild.bat"))
+
+
+@pytest.mark.skipif(platform.system() not in ["Windows"], reason="Requires Windows")
+def test_vcvars_generator_skip():
+    """
+    tools.microsoft.msbuild:installation_path=disabled avoids creation of conanvcvars.bat
+    """
+    client = TestClient()
+    client.save({"conanfile.py": GenConanfile().with_generator("VCVars")
+                                               .with_settings("os", "compiler",
+                                                              "arch", "build_type")})
+    client.run('install . -c tools.microsoft.msbuild:installation_path=disabled')
+    assert not os.path.exists(os.path.join(client.current_folder, "conanvcvars.bat"))
 
 
 @pytest.mark.skipif(platform.system() not in ["Windows"], reason="Requires Windows")

